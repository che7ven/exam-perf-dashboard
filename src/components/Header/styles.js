const styles = (theme) => ({
  card: {
    minWidth: 93,
  },
  center: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    flexDirection: "column",
  },
  hr: {
    width: "60%",
    margin: 10,
  },
  mainUser: {
    width: 45,
    height: 45,
    borderRadius: 23,
  },
  nav: {
    top: 0,
    height: 60,
    minWidth: "100vw",
    display: "flex",
    alignItems: "center",
    justifyContent: "space-between",
    position: "fixed",
    background: "#041F41",
    color: "azure",
    zIndex: 9999,
    boxShadow: "3px 5px 15px 3px #9a9898",
    [theme.breakpoints.up("sm")]: {
      fontSize: 16,
      height: 64,
    },
  },
  notifyIcon: {
    width: 24,
    height: 24,
  },
  optionHolder: {
    marginRight: 22,
    [theme.breakpoints.up("sm")]: {
      marginRight: 4,
    },
    [theme.breakpoints.up("md")]: {
      marginRight: 2,
    },
    [theme.breakpoints.up("lg")]: {
      marginRight: 18,
    },
  },
  subText: {
    fontSize: 12,
    color: "#0071ce",
  },
  userName: {
    background: "#d8d8d8",
    width: 24,
    height: 22,
    borderRadius: 16,
    fontSize: 16,
    fontWeight: 500,
    color: "#000000",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    [theme.breakpoints.up("lg")]: {
      width: 32,
      height: 32,
    },
  },
  btnStyle: {
    width: "100%",
    height: 48,
    fontSize: 14,
    color: "#282828",
    textTransform: "none !important",
    justifyContent: "left",
    "&:hover": {
      backgroundColor: "#f2f2f2",
    },
  },
});

export default styles;
