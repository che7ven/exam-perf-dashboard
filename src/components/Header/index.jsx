import React, { useState } from "react";
import PropTypes from "prop-types";
import { useNavigate } from "react-router-dom";
import {
  Card,
  IconButton,
  Popover,
  Typography,
  Button,
  withStyles,
} from "@material-ui/core";
import { getOrigin } from "../../utils/constants";
import styles from "./styles";

const Header = ({ classes, user }) => {
  const navigate = useNavigate();

  const [userProfile, setUserProfile] = useState(null);

  const openUserProfile = (event) => {
    setUserProfile(event.currentTarget);
  };

  const closeUserProfile = () => {
    setUserProfile(null);
  };

  const userOpen = Boolean(userProfile);
  const userId = userOpen ? "user-detail" : undefined;

  const logout = () => {
    navigate("/", { state: "logOff" });
  };

  return (
    <header>
      <nav className={classes.nav}>
        <Typography variant="h5" style={{ paddingLeft: 8 }}>
          Exam Dashboard
        </Typography>
        <div className={classes.optionHolder}>
          {user && (
            <IconButton
              onClick={openUserProfile}
              id={userId}
              aria-label="profile"
              data-testid={"profile"}
            >
              <div className={classes.userName}>
                {user && user?.charAt(0)?.toUpperCase()}
              </div>
            </IconButton>
          )}
          <Popover
            id={userId}
            open={userOpen}
            anchorEl={userProfile}
            onClose={closeUserProfile}
            anchorOrigin={getOrigin("bottom", "left")}
            transformOrigin={getOrigin("top", "center")}
            data-testid={"profile-pop"}
          >
            <Card
              className={`${classes.typography} ${classes.center} ${classes.card}`}
              role="dialog"
            >
              <Button
                id="logout"
                className={classes.btnStyle}
                onClick={logout}
                data-testid={"logout"}
              >
                LOGOUT
              </Button>
            </Card>
          </Popover>
        </div>
      </nav>
    </header>
  );
};

Header.defaultProps = {
  classes: {},
  type: "",
};

Header.propTypes = {
  classes: PropTypes.object,
  user: PropTypes.string,
};

export default withStyles(styles)(Header);
