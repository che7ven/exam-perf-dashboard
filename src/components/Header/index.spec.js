import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import Header from "../Header";

const mockedUseNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("Header", () => {
  let props;

  beforeEach(() => {
    props = { user: "chevy" };
  });

  it("should take a snapshot", () => {
    const { asFragment } = render(<Header {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should call event handler on user profile click", () => {
    render(<Header {...props} />);
    const tab = screen.getByTestId("profile");

    fireEvent.click(tab);
    expect(tab.id).toBe("user-detail");
    expect(screen.queryByTestId("logout")).not.toBeNull();

    const outerEl = screen.getByTestId("profile-pop").firstChild;
    fireEvent.click(outerEl);
  });

  it("should navigate to login page on logout click", () => {
    const { getByTestId } = render(<Header {...props} />);
    const profile = getByTestId("profile");
    fireEvent.click(profile);

    const logout = getByTestId("logout");
    fireEvent.click(logout);

    expect(mockedUseNavigate).toHaveBeenNthCalledWith(1, "/", {
      state: "logOff",
    });
  });
});
