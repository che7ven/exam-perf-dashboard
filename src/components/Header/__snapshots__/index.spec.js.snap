// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should take a snapshot 1`] = `
<DocumentFragment>
  <header>
    <nav
      class="Header-nav-5"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5"
        style="padding-left: 8px;"
      >
        Exam Dashboard
      </h5>
      <div
        class="Header-optionHolder-7"
      >
        <button
          aria-label="profile"
          class="MuiButtonBase-root MuiIconButton-root"
          data-testid="profile"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <div
              class="Header-userName-9"
            >
              C
            </div>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </nav>
  </header>
</DocumentFragment>
`;
